runtime/lua/vim/lsp/sync.lua

```diff
@@ -0,0 +1,384 @@
+--  offset of b is 3 since 𐐀 is represented using two code units in UTF-16.
+--
+--  To ensure that both client and server split the string into the same line
+--  representation the protocol specifies the following end-of-line sequences: ‘\n’, ‘\r\n’ and ‘\r’.
+--
+--  Positions are line end character agnostic. So you can not specify a position that
+--  denotes \r|\n or \n| where | represents the character offset. This means *no* defining
+--  a range than ends on the same line after a terminating character
+--
+-- Generic warnings about byte level changes in neovim. Many apparently "single"
+-- operations in on_lines callbacks are actually multiple operations.
+--
+--  Join operation (2 operations):
+--  * extends line 1 with the contents of line 2
+--  * deletes line 2
+--
+--  test 1    test 1 test 2    test 1 test 2
+--  test 2 -> test 2        -> test 3
+--  test 3    test 3
+--
+--  Deleting (and undoing) two middle lines (1 operation):
+--
+--  test 1    test 1
+--  test 2 -> test 4
+--  test 3
+--  test 4
+--
+--  Deleting partial lines (5 operations) deleting between asterisks below:
+--
+--  test *1   test *    test *     test *    test *4    test *4*
+--  test 2 -> test 2 -> test *4 -> *4     -> *4      ->
+--  test 3    test 3
+--  test *4   test 4
+
+local M = {}
+
+-- local string.byte, unclear if this is necessary for JIT compilation
+local str_byte = string.byte
+local vim = vim
+local math = math
+
+---@private
+-- Given a line, byte idx, and offset_encoding convert to the
+-- utf-8, utf-16, or utf-32 index.
+---@param line string the line to index into
+---@param byte integer the byte idx
+---@param offset_encoding string utf-8|utf-16|utf-32|nil (default: utf-8)
+--@returns integer the utf idx for the given encoding
+function M.byte_to_utf(line, byte, offset_encoding)
+  -- convert to 0 based indexing for str_utfindex
+  byte = byte - 1
+
+  local utf_idx
+  local _
+  -- Convert the byte range to utf-{8,16,32} and convert 1-based (lua) indexing to 0-based
+  if offset_encoding == 'utf-16' then
+    _, utf_idx = vim.str_utfindex(line, byte)
+  elseif offset_encoding == 'utf-32' then
+    utf_idx, _ = vim.str_utfindex(line, byte)
+  else
+    utf_idx = byte
+  end
+
+  -- convert to 1 based indexing
+  return utf_idx + 1
+end
+
+---@private
+-- Given a line, byte idx, alignment, and offset_encoding convert to the aligned
+-- utf-8 index and either the utf-16, or utf-32 index.
+---@param line string the line to index into
+---@param byte integer the byte idx
+---@param align string when dealing with multibyte characters,
+--        to choose the start of the current character or the beginning of the next.
+--        Used for incremental sync for start/end range respectively
+---@param offset_encoding string utf-8|utf-16|utf-32|nil (default: utf-8)
+---@returns table<string, int> byte_idx and char_idx of first change position
+function M.align_position(line, byte, align, offset_encoding)
```

Comment:
Only `compute_diff` is used by the other modules. Should we keep the other functions local? (Or rename `sync` to `_sync` to indicate that it's private?)

Reply:
Yes, I'll make everything private. Originally I kept these public because I was going to write tests foreach one, but I ended up writing end to end tests.